File: scratch.py
def length_of_longest_substring(s) -> int:
    char_map = {}
    left_pointer = 0
    result = 0

    for right_pointer in range(0, len(s), 1):
        right_char = s[right_pointer]

        if right_char in char_map:
            left_pointer = max(left_pointer, char_map[right_char] + 1)

        char_map[right_char] = right_pointer
        result = max(result, right_pointer - left_pointer + 1)

    return result


print(length_of_longest_substring("abcabcbb"))  # output: 3
print(length_of_longest_substring("abcabcbbabcde"))  # output: 5
print(length_of_longest_substring("bbbbb"))  # output: 1
print(length_of_longest_substring("pwwkew"))  # output: 3

File: scratch.js
function lengthOfLongestSubstring(s) {
    let charMap = {}
    let result = 0
    let leftP = 0

    for (let rightP = 0; rightP < s.length; rightP += 1) {
        let rightChar = s[rightP]

        if (rightChar in charMap) {
            leftP = Math.max(leftP, charMap[rightChar] + 1)
        }

        charMap[rightChar] = rightP
        result = Math.max(result, rightP - leftP + 1)
    }

    return result
}


console.log(lengthOfLongestSubstring("abcabcbb"));  // output: 3
console.log(lengthOfLongestSubstring("abcabcbbabcde"));  // output: 5
console.log(lengthOfLongestSubstring("bbbbb"));  // output: 1
console.log(lengthOfLongestSubstring("pwwkew"));  // output: 3
